File: backend/services/financial/transaction_search.py
"""Boolean transaction search; same grammar as the investigator's table."""
import re
from services.financial.money import get_currency

FIELDS = dict(from_='from_name', to='to_name', category='category', description='description', reference='bank_reference', date='ordering_date', currency='currency', account='account_label')
FIELDS['from'] = FIELDS.pop('from_')


def transaction_search(query, mode='text'):
    tokens = re.findall(r'(?:[a-zA-Z]+:)?"(?:\\.|[^"\\])*"|\(|\)|[^\s()"]+', query)
    index = 0
    def primary():
        nonlocal index
        token = tokens[index] if index < len(tokens) else None
        index += 1
        if token is None or token in ('AND', 'OR', ')'):
            raise ValueError('Enter a term after each operator.')
        if token == 'NOT':
            return ('NOT', primary())
        if token == '(':
            value = either()
            if index >= len(tokens) or tokens[index] != ')':
                raise ValueError('Close each opening parenthesis.')
            index += 1
            return value
        match = re.fullmatch(r'([a-zA-Z]+):(.*)', token)
        field, term = (match[1].lower(), match[2]) if match else (None, token)
        if field and field not in FIELDS and field != 'amount':
            raise ValueError('Unknown search field.')
        if not term:
            raise ValueError('Enter a value after the field name.')
        if term.startswith('"'):
            term = re.sub(r'\\(.)', r'\1', term[1:-1])
        if not term.strip():
            raise ValueError('Enter text inside quoted phrases.')
        return ('term', term.lower(), field)
    def both():
        nonlocal index
        value = primary()
        while index < len(tokens) and tokens[index] not in ('OR', ')'):
            if tokens[index] == 'AND':
                index += 1
            value = ('AND', value, primary())
        return value
    def either():
        nonlocal index
        value = both()
        while index < len(tokens) and tokens[index] == 'OR':
            index += 1
            value = ('OR', value, both())
        return value
    if not query.strip():
        tree = None
    elif mode == 'boolean':
        if re.sub(r'\s', '', ''.join(tokens)) != re.sub(r'\s', '', query):
            raise ValueError('Close each quoted phrase before searching.')
        tree = either()
        if index != len(tokens):
            raise ValueError('Remove the unmatched closing parenthesis.')
    else:
        tree = ('term', query.strip().lower(), None)
    def matches(row):
        values = {name: str(row.get(field) or '') for name, field in FIELDS.items()}
        values['category'] = values['category'] or 'Uncategorized'
        scale = get_currency(row['currency']).exponent
        minor = str(row['amount_minor']).zfill(scale + 1)
        values['amount'] = minor if not scale else minor[:-scale] + '.' + minor[-scale:]
        text = ' '.join([*values.values(), *[str(row.get(key) or '') for key in ('ref_id', 'key', 'counterparty_raw', 'account_holder', 'account_id')]]).lower()
        def evaluate(node):
            if node[0] == 'term':
                return node[1] in (values[node[2]].lower() if node[2] else text)
            if node[0] == 'NOT':
                return not evaluate(node[1])
            return (evaluate(node[1]) and evaluate(node[2])) if node[0] == 'AND' else (evaluate(node[1]) or evaluate(node[2]))
        return tree is None or evaluate(tree)
    return matches

File: frontend_v2/src/features/financial/lib/transaction-search.ts
import { currencyMinorUnits } from "./ledger-format"
import type { LedgerTransaction } from "../api"
import { correctionMoney } from "./correction-contract"

type Node =
  | { term: string; field?: string }
  | { op: "NOT"; value: Node }
  | { op: "AND" | "OR"; left: Node; right: Node }
const fields: Record<string, keyof LedgerTransaction> = {
  from: "from_name",
  to: "to_name",
  category: "category",
  description: "description",
  reference: "bank_reference",
  date: "ordering_date",
  currency: "currency",
  account: "account_label",
}
export function parseTransactionSearch(query: string): Node | null {
  if (!query.trim()) return null
  const tokens =
    query.match(/(?:[a-zA-Z]+:)?"(?:\\.|[^"\\])*"|\(|\)|[^\s()"]+/g) ?? []
  if (tokens.join("").replace(/\s/g, "") !== query.replace(/\s/g, ""))
    throw Error("Close each quoted phrase before searching.")
  let index = 0
  const primary = (): Node => {
    const token = tokens[index++]
    if (!token || ["AND", "OR", ")"].includes(token))
      throw Error("Enter a term after each operator.")
    if (token === "NOT") return { op: "NOT", value: primary() }
    if (token === "(") {
      const value = or()
      if (tokens[index++] !== ")")
        throw Error("Close each opening parenthesis.")
      return value
    }
    const match = token.match(/^([a-zA-Z]+):(.*)$/)
    const field = match?.[1].toLowerCase()
    if (field && !(field in fields) && field !== "amount")
      throw Error(`Unknown search field: ${field}.`)
    let term = match ? match[2] : token
    if (!term) throw Error("Enter a value after the field name.")
    if (term.startsWith('"')) term = term.slice(1, -1).replace(/\\(.)/g, "$1")
    if (!term.trim()) throw Error("Enter text inside quoted phrases.")
    return { term: term.toLowerCase(), ...(field ? { field } : {}) }
  }
  const and = (): Node => {
    let value = primary()
    while (index < tokens.length && !["OR", ")"].includes(tokens[index])) {
      if (tokens[index] === "AND") index++
      value = { op: "AND", left: value, right: primary() }
    }
    return value
  }
  const or = (): Node => {
    let value = and()
    while (tokens[index] === "OR") {
      index++
      value = { op: "OR", left: value, right: and() }
    }
    return value
  }
  const result = or()
  if (index !== tokens.length)
    throw Error("Remove the unmatched closing parenthesis.")
  return result
}
export function transactionSearch(query: string, mode: string = "text") {
  let tree: Node | null
  try {
    tree =
      mode === "boolean"
        ? parseTransactionSearch(query)
        : query.trim()
          ? { term: query.trim().toLowerCase() }
          : null
  } catch (error) {
    return { error: (error as Error).message, matches: () => false }
  }
  if (!tree) return { error: null, matches: () => true }
  const matches = (row: LedgerTransaction) => {
    const values: Record<string, string> = Object.fromEntries(
      Object.entries(fields).map(([name, field]) => [
        name,
        String(row[field] ?? ""),
      ])
    )
    values.category ||= "Uncategorized"
    values.amount = correctionMoney(
      String(row.amount_minor),
      row.currency
    ).split(" ")[0]
    const all = [
      ...Object.values(values),
      row.ref_id,
      row.key,
      row.counterparty_raw,
      row.account_holder,
      row.account_id,
    ]
      .filter(Boolean)
      .join(" ")
      .toLowerCase()
    const evaluate = (node: Node): boolean =>
      "term" in node
        ? (node.field ? values[node.field].toLowerCase() : all).includes(
            node.term
          )
        : node.op === "NOT"
          ? !evaluate(node.value)
          : node.op === "AND"
            ? evaluate(node.left) && evaluate(node.right)
            : evaluate(node.left) || evaluate(node.right)
    return !tree || evaluate(tree)
  }
  return { error: null, matches }
}

export function compareDisplayedAmounts(
  a: LedgerTransaction,
  b: LedgerTransaction
) {
  const leftScale = currencyMinorUnits(a.currency) ?? 0
  const rightScale = currencyMinorUnits(b.currency) ?? 0
  const left = BigInt(a.amount_minor) * 10n ** BigInt(rightScale)
  const right = BigInt(b.amount_minor) * 10n ** BigInt(leftScale)
  return left < right ? -1 : left > right ? 1 : 0
}
